src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java

```diff
@@ -0,0 +1,258 @@
+     * Asynchronously update the progress for learning goals linked to the given learning object (for the specified participant)
+     * @param learningObject The learning object for which to fetch the learning goals
+     * @param participant The participant (user or team) for which to update the progress
+     */
+    @Async
+    public void updateProgressByLearningObjectAsync(ILearningObject learningObject, @NotNull Participant participant) {
+        SecurityUtils.setAuthorizationObject(); // required for async
+        updateProgressByLearningObject(learningObject, participant.getParticipants());
+    }
+
+    /**
+     * Asynchronously update the progress for the learning goals linked to the given learning object (for all students in the course)
+     * @param learningObject The learning object for which to fetch the learning goals
+     */
+    @Async
+    public void updateProgressByLearningObjectAsync(ILearningObject learningObject) {
+        SecurityUtils.setAuthorizationObject(); // required for async
+        Course course;
+        if (learningObject instanceof Exercise exercise) {
+            course = exercise.getCourseViaExerciseGroupOrCourseMember();
+        }
+        else if (learningObject instanceof LectureUnit lectureUnit) {
+            course = lectureUnit.getLecture().getCourse();
+        }
+        else {
+            throw new IllegalArgumentException("Learning object must be either LectureUnit or Exercise");
+        }
+        updateProgressByLearningObject(learningObject, new HashSet<>(userRepository.getStudents(course)));
+    }
+
+    /**
+     * Asynchronously update the existing progress for a specific learning goal
+     * @param learningGoal The learning goal for which to update all existing student progress
+     */
+    @Async
+    public void updateProgressByLearningGoalAsync(LearningGoal learningGoal) {
+        SecurityUtils.setAuthorizationObject(); // required for async
+        learningGoalProgressRepository.findAllByLearningGoalId(learningGoal.getId()).stream().map(LearningGoalProgress::getUser).forEach(user -> {
+            updateLearningGoalProgress(learningGoal.getId(), user);
+        });
+    }
+
+    /**
+     * Update the existing progress for a specific user in a course
+     * @param user The user for whom to update the existing learning goal progress
+     * @param course The course for which to fetch the learning goals from
+     * @return All learning goals of the course with the updated progress for the user
+     */
+    public Set<LearningGoal> getLearningGoalsAndUpdateProgressByUserInCourse(User user, Course course) {
+        return learningGoalRepository.findAllForCourse(course.getId()).stream().peek(learningGoal -> {
+            var updatedProgress = updateLearningGoalProgress(learningGoal.getId(), user);
+            if (updatedProgress != null) {
+                learningGoal.setUserProgress(Set.of(updatedProgress));
+            }
+        }).collect(Collectors.toSet());
```

Comment:
Please avoid using peek. From the [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer-): "API Note: This method exists mainly to support debugging, where you want to see the elements as they flow past a certain point in a pipeline."
```suggestion
        var learningGoals = learningGoalRepository.findAllForCourse(course.getId());
        learningGloals.forEach(learningGoal -> {
            var updatedProgress = updateLearningGoalProgress(learningGoal.getId(), user);
            if (updatedProgress != null) {
                learningGoal.setUserProgress(Set.of(updatedProgress));
            }
        });
        return learningGoals;
```